lock-api/src/main/java/com/palantir/lock/client/LeasingTimelockClient.java

```diff
@@ -0,0 +1,160 @@
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.lock.v2.WaitForLocksResponse;
+import com.palantir.timestamp.TimestampRange;
+
+public final class LeasingTimelockClient implements TimelockService {
+    private final TimelockRpcClient delegate;
+
+    private LeasingTimelockClient(TimelockRpcClient timelockRpcClient) {
+        this.delegate = timelockRpcClient;
+    }
+
+    public static LeasingTimelockClient create(TimelockRpcClient timelockRpcClient) {
+        return new LeasingTimelockClient(timelockRpcClient);
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return delegate.getFreshTimestamp();
+    }
+
+    @Override
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return delegate.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @Override
+    public LockImmutableTimestampResponse lockImmutableTimestamp(IdentifiedTimeLockRequest request) {
+        return delegate.lockImmutableTimestamp(request);
+    }
+
+    @Override
+    public StartAtlasDbTransactionResponse startAtlasDbTransaction(IdentifiedTimeLockRequest request) {
+        return delegate.startAtlasDbTransaction(request);
+    }
+
+    @Override
+    public StartIdentifiedAtlasDbTransactionResponse startIdentifiedAtlasDbTransaction(
+            StartIdentifiedAtlasDbTransactionRequest request) {
+        LeasableStartIdentifiedAtlasDbTransactionResponse leasableResponse =
+                delegate.leasableStartIdentifiedAtlasDbTransaction(request);
+
+        StartIdentifiedAtlasDbTransactionResponse response = leasableResponse.getStartTransactionResponse();
+        Lease lease = leasableResponse.getLease();
+        LeasedLockToken leasedLockToken = LeasedLockToken.of(response.immutableTimestamp().getLock(), lease);
+        long immutableTs = response.immutableTimestamp().getImmutableTimestamp();
+
+        return ImmutableStartIdentifiedAtlasDbTransactionResponse.of(
+                ImmutableLockImmutableTimestampResponse.of(immutableTs, leasedLockToken),
+                response.startTimestampAndPartition());
+    }
+
+    @Override
+    public long getImmutableTimestamp() {
+        return delegate.getImmutableTimestamp();
+    }
+
+    @Override
+    public LockResponse lock(LockRequest request) {
+        LeasableLockResponse leasableResponse = delegate.leasableLock(request);
+
+        LockResponse lockResponse = leasableResponse.getLockResponse();
+        Lease lease = leasableResponse.getLease();
+        LeasedLockToken leasedLockToken = LeasedLockToken.of(lockResponse.getToken(), lease);
+
+        return ImmutableLockResponse.of(Optional.of(leasedLockToken));
+    }
+
+    @Override
+    public WaitForLocksResponse waitForLocks(WaitForLocksRequest request) {
+        return delegate.waitForLocks(request);
+    }
+
+    @Override
+    public Set<LockToken> refreshLockLeases(Set<LockToken> tokens) {
```

Comment:
better call this 'uncastedTokens'; the variable name 'tokens' doesn't play well with 'allTokens', doesn't really describe why they're different.